### ISSUE TITLE
Components are unmounted / never mounted if they return null

### ISSUE THREAD

**Issue by spiffytech** [forgojs/forgo#39]
#37 got me curious to seek clarity on what Forgo does when a component returns null. I see that Forgo unmounts the component if it returns null, throwing away the component's state ([sandbox](https://codesandbox.io/s/forgo-null-return-myyjy)).

This makes sense given the explanation that [component state is stored on its DOM node](https://github.com/forgojs/forgo/issues/37#issuecomment-1019411349), but is unexpected as an application developer. The decision not to display any HTML seems unrelated to the decision to discard component state. For example, if a component wants to [return null until a resource has loaded](https://codesandbox.io/s/forgo-null-return-forked-wjx1q?file=/src/index.tsx) that seems to not work right now.

If Forgo bases all its internal bookkeeping assuming a component always corresponds to a DOM node, does that mean this behavior would be pretty involved to change? I think even `Fragment`s are implemented by selecting the first DOM node they contain, right?

What are your thoughts here?

**Comment by jeswin** [comment]
Yes, correct. Components store their state in the element that DOM element(s) that finally get rendered. When/if the element is deleted, the component also gets unmounted. This is what allows it to skip the vdom.

As you wrote, this is so fundamental that it's impossible to change. 

I think it's a compromise, but perhaps it needs to be made clearer in the docs (in an advanced section).  Something like -  "Forgo maintains component state in DOM elements as opposed to having a vdom like react. Components which don't render an element are unmounted, so if you want to keep it around render an empty element. See example below."

Aside from avoiding the vdom, another benefit which isn't implemented yet is that one could use jQuery to take out an element and corresponding components will unmount.

**Comment by jeswin** [comment]
> Does it make sense to allow a component to return null?

I'm thinking we should allow this. Stateless components which have nothing to display might render a null. 

```jsx
<div>
{ 
   number.map(num => <DisplayTickMarkIfEven value={num} />
}
</div>
```

**Comment by spiffytech** [comment]
Sounds good, I'll add a note about doing that in the readme.

**Comment by spiffytech** [comment]
Fun fact: HTML comments count as DOM nodes. You can add/remove/retrieve them, and add properties onto them. It seems pretty similar to using text nodes.

At some point I'm going to take a crack at handling null renders by rendering an HTML comment that we can put the component state onto. That should let stateful components recover from a null render, without polluting the DOM with stuff like `div`s that have meaning in the document.

**Comment by jeswin** [comment]
This is a good idea! Go for it.
